### ISSUE TITLE
FR: Volume Preserving Real NVP Variant (GIN)

### ISSUE THREAD

**Issue by robsalomone** [pyro-ppl/pyro#2747]
Hi all, 

I noticed that Real-NVP has an option for "Volume Preserving", but this just omits the scale factors, and performs the NICE transform. Would it be possible to implement a more powerful volume preserving transformation (I need it for a specific purpose) by simply modifying Real-NVP so the final scale variable is the negative sum of the others?

The very simple idea can be found in this ICLR2020 paper: https://arxiv.org/abs/2001.04872

Similarly, it should be possible to do this for IAF.

It should be easy for someone who is familiar with the code already.

**Comment by eb8680** [comment]
cc @stefanwebb

**Comment by stefanwebb** [comment]
Hi @robsalomone, I'm the guy responsible for the existing normalizing flow code in Pyro! Yes, I think it should be quite simple to make this adjustment. Let me read the paper and report back - I can't promise a timeline but will look into it soon

**Comment by robsalomone** [comment]
Thanks @stefanwebb! Reading the paper isn't (entirely) necessary, as the main idea in essence is quite simple: for any NF such that the determinant is the product of some diagonal entries, one can simply enforce that the last diagonal element is the negative sum of the others. Then, the determinant becomes one and the map is volume preserving.  

In RealNVP or IAF/MAF, this is equivalent to just changing the last scale factor appropriately. Thus, it does not have its own parametrisation anymore, and is simply a linear function of the other $s$ functions.

**Comment by robsalomone** [comment]
Of course, you may be interested in reading the paper for why its a good idea in general. However, I confess the reason I would like volume preservation is a very specific one that isn't the one mentioned in the paper!
